src/rqt_graph/dotcode.py

```diff
@@ -498,41 +654,98 @@ def generate_dotgraph(self,
+                else:
+                    namespace = '/'.join(n.strip().split('/')[:cluster_namespaces_level+1])
+                if namespace not in namespace_clusters:
+                    print("Namespace '"+namespace+"' not found.")
+                self._add_topic_node(n, dotcode_factory=dotcode_factory, dotgraph=namespace_clusters[namespace], quiet=quiet)
+             else:
+                self._add_topic_node(n, dotcode_factory=dotcode_factory, dotgraph=dotgraph, quiet=quiet)
+
+        for n in [action_prefix + ACTION_TOPICS_SUFFIX for (action_prefix, _) in action_nodes.items()]:
             # cluster topics with same namespace
             if (cluster_namespaces_level > 0 and
-                str(n).count('/') > 1 and
-                len(str(n).split('/')[1]) > 0):
-                namespace = str(n).split('/')[1]
+                str(n).strip().count('/') > 1 and
+                len(str(n).strip().split('/')[1]) > 0):
+                if (n.strip().count('/') <= cluster_namespaces_level):
+                    namespace = str('/'.join(n.strip().split('/')[:-1]))
+                else:
+                    namespace = '/'.join(n.strip().split('/')[:cluster_namespaces_level+1])
                 if namespace not in namespace_clusters:
-                    namespace_clusters[namespace] = dotcode_factory.add_subgraph_to_graph(dotgraph, namespace, rank=rank, rankdir=orientation, simplify=simplify)
-                self._add_topic_node(n, dotcode_factory=dotcode_factory, dotgraph=namespace_clusters[namespace], quiet=quiet)
+                    print("Namespace '"+namespace+"' not found.")
+                self._add_topic_node_group('n'+n, dotcode_factory=dotcode_factory, dotgraph=namespace_clusters[namespace], quiet=quiet)
             else:
-                self._add_topic_node(n, dotcode_factory=dotcode_factory, dotgraph=dotgraph, quiet=quiet)
+                self._add_topic_node_group('n'+n, dotcode_factory=dotcode_factory, dotgraph=dotgraph, quiet=quiet)
+
+        for n in [image_prefix + IMAGE_TOPICS_SUFFIX for (image_prefix, _) in image_nodes.items()]:
+            # cluster topics with same namespace
+            if (cluster_namespaces_level > 0 and
+                str(n).strip().count('/') > 1 and
+                len(str(n).strip().split('/')[1]) > 0):
+                if (n.strip().count('/') <= cluster_namespaces_level):
+                    namespace = str('/'.join(n.strip().split('/')[:-1]))
+                else:
+                    namespace = '/'.join(n.strip().split('/')[:cluster_namespaces_level+1])
+                if namespace not in namespace_clusters:
+                    print("Namespace '"+namespace+"' not found.")
+                self._add_topic_node_group('n'+n, dotcode_factory=dotcode_factory, dotgraph=namespace_clusters[namespace], quiet=quiet)
+            else:
+                self._add_topic_node_group('n'+n, dotcode_factory=dotcode_factory, dotgraph=dotgraph, quiet=quiet)
```

Comment:
These three loops look partially very similar. Maybe the code would get shorter by refactoring some of this into a helper function?

*Update:* and a fourth time below.